cartographer/mapping_3d/pose_graph/landmark_cost_function.h

```diff
@@ -0,0 +1,124 @@
+ * limitations under the License.
+ */
+
+#ifndef CARTOGRAPHER_MAPPING_3D_POSE_GRAPH_LANDMARK_COST_FUNCTION_H_
+#define CARTOGRAPHER_MAPPING_3D_POSE_GRAPH_LANDMARK_COST_FUNCTION_H_
+
+#include "Eigen/Core"
+#include "Eigen/Geometry"
+#include "cartographer/mapping/pose_graph.h"
+#include "cartographer/mapping_3d/pose_graph/spa_cost_function.h"
+#include "cartographer/transform/rigid_transform.h"
+#include "cartographer/transform/transform.h"
+#include "ceres/ceres.h"
+#include "ceres/jet.h"
+
+namespace cartographer {
+namespace mapping_3d {
+namespace pose_graph {
+
+// Cost function measuring weighted error between the observed relative pose
+// given by sensor and relative pose computed from the linearly interpolated
+// pose of the robot at the moment of observation.
+class LandmarkCostFunction {
+ public:
+  using LandmarkObservation =
+      mapping::PoseGraph::LandmarkNode::LandmarkObservation;
+
+  static ceres::CostFunction* CreateAutoDiffCostFunction(
+      const LandmarkObservation& observation, common::Time prev_node_time,
+      common::Time next_node_time) {
+    return new ceres::AutoDiffCostFunction<
+        LandmarkCostFunction, 6 /* residuals */,
+        3 /* previous node translation variables */,
+        4 /* previous node rotation variables */,
+        3 /* next node translation variables */,
+        4 /* next node rotation variables */,
+        3 /* landmark translation variables */,
+        4 /* landmark rotation variables */>(
+        new LandmarkCostFunction(observation, prev_node_time, next_node_time));
+  }
+
+  template <typename T>
+  bool operator()(const T* const prev_node_rotation,
+                  const T* const prev_node_translation,
+                  const T* const next_node_rotation,
+                  const T* const next_node_translation,
+                  const T* const landmark_rotation,
+                  const T* const landmark_translation, T* const e) const {
+    const T interpolated_pose_translation[3] = {
+        prev_node_translation[0] +
+            interpolation_parameter_ *
+                (next_node_translation[0] - prev_node_translation[0]),
+        prev_node_translation[1] +
+            interpolation_parameter_ *
+                (next_node_translation[1] - prev_node_translation[1]),
+        prev_node_translation[2] +
+            interpolation_parameter_ *
+                (next_node_translation[2] - prev_node_translation[2])};
+
+    const Eigen::Quaternion<T> prev_quaternion(
+        prev_node_rotation[0], prev_node_rotation[1], prev_node_rotation[2],
+        prev_node_rotation[3]);
+    const Eigen::Quaternion<T> next_quaternion(
+        next_node_rotation[0], next_node_rotation[1], next_node_rotation[2],
+        next_node_rotation[3]);
+
+    const Eigen::Quaternion<T> interpolated_quaternion =
+        prev_quaternion.slerp(T(interpolation_parameter_), next_quaternion);
+    const T interpolated_pose_rotation[4] = {
```

Comment:
if the pose interpolation is commonly used, it could go into a helper function.

Reply:
It is not commonly used yet. However, ComputeUnscaledError/ComputeScaledError should go into some lib. Not in this PR though.